File: JAVASCRIPT/RECURSION/mergeSort.js
const mergeSort = arr => {
  // dividing the array until getting one element arrays ************************
  if (arr.length <= 1) {
    return arr;
  }
  
  const mid = Math.floor(arr.length / 2);
  
  const left = mergeSort(arr.slice(0, mid));
  const right = mergeSort(arr.slice(mid));
  
  // sorting the arrays ********************************************************
  const sorted = [];
  let i = 0;
  let j = 0;
  console.log('left array: ', left)
  console.log('right array: ', right)
  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]){
        sorted.push(left[i])
        i++
    } else {
        sorted.push(right[j])
        j++
    }
    console.log('sorted: ', sorted)
  }
  // concatenating the sorted array with remaining parts of left & right arrays *******
  return sorted.concat(left.slice(i)).concat(right.slice(j))
};

console.log(mergeSort([4, 1, 3, 5, 2]))

File: Python/Algorithm/merge-sort.py
import random
from PIL import Image

def show_graph():
    img = Image.open('O(n-log-n).png')
    img.show()

def merge_sort(arr, depth=0):
    '''recursive algorithm, variable 'depth' is used to show the 
    recursion mechanism. Sorting logic after base case is
    split, compare, merge (extend);
     because unlike other sorting algorithm is not 
    sorting in place'''
    
    indent = ' ' * depth
    print(f'{indent}merge_sort called with: {arr}')
    # base case
    if len(arr) <= 1:
        # print(f'{indent}Base case reached: returning {arr}')
        return arr

    mid = len(arr) // 2
    print(f'{indent}Splitting: left = {arr[:mid]}, right = {arr[mid:]}')
    # print(f'mid: {mid}')
    # recursive calls continue until base case
    left = merge_sort(arr[:mid], depth + 2)
    # print(f'left: {left}')
    right = merge_sort(arr[mid:], depth + 2)
    # print(f'right: {right}')

    # after recursive calls end the call stack starts releasing 
    # left and right and compares them in the while loop
    sorted_list = []
    i = 0
    j = 0

###########while loop scope######################################
    while i < len(left) and j < len(right):
        # loop logic after split: append lower then extend with 
        # the other one
        if left[i] <= right[j]: # compares single values
            sorted_list.append(left[i]) # appends them
            print(f'sorted list append left: {left[i]}')
            i += 1
        else:
            sorted_list.append(right[j])
            print(f'sorted list append right: {right[j]}')
            j += 1
###################################################################

    sorted_list.extend(left[i:])
    print(f'{indent}Sorted list extend left: {left[i:]}')
    sorted_list.extend(right[j:])
    print(f'{indent}Sorted list extend right: {right[j:]}')

    print(f'{indent}Returning merged: {sorted_list}')
    return sorted_list
    # after the return the function keeps rolling because the call stack
    # is still returning calls !!
test = merge_sort([random.randint(1, 100) for _ in range(4)])
print(test)

show_graph()